go/pkg/cas/upload.go

```diff
@@ -358,24 +388,67 @@ func (u *uploader) visitDir(ctx context.Context, absPath string) (*repb.Director
 
 	wgChildren.Wait()
 	if subErr != nil {
-		return nil, errors.Wrapf(subErr, "failed to read the directory %q entirely", absPath)
+		return nil, errors.Wrapf(subErr, "failed to read the directory %q entirely", key.AbsPath)
 	}
 
-	item := uploadItemFromDirMsg(absPath, dir)
+	item := uploadItemFromDirMsg(key.AbsPath, dir)
 	if err := u.scheduleCheck(ctx, item); err != nil {
 		return nil, err
 	}
 	return &repb.DirectoryNode{
-		Name:   filepath.Base(absPath),
+		Name:   filepath.Base(key.AbsPath),
 		Digest: item.Digest,
 	}, nil
 }
 
-// visitSymlink converts a symlink to a SymlinkNode and schedules visitation
+// visitSymlink converts a symlink to a directory node and schedules visitation
 // of the target file.
-func (u *uploader) visitSymlink(ctx context.Context, absPath string) (*repb.SymlinkNode, error) {
-	// TODO(nodir): implement
-	panic("not supported")
+// If key.PreserveSymlinks is true, then returns a SymlinkNode, otherwise
+// returns the directory node of the target file.
+func (u *uploader) visitSymlink(ctx context.Context, key fsCacheKey) (proto.Message, error) {
+	target, err := os.Readlink(key.AbsPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "os.ReadLink")
+	}
+
+	// Determine absolute and relative paths of the target.
+	var absTarget, relTarget string
```

Comment:
I just remembered when checking this `relTarget`: generally action inputs are restricted under the execution root (or input root) of an action. If we're preserving symlinks, and their targets are outside of the input root, they'll be effectively dangling. When creating the input (the `Tree`), the paths also have to be relative to the execution root.

I think we should have a ClientConfig that specifies an "ExecRoot", and if specified, we should be err-ing UploadInputs outside of the root on `startProcessing`, and failing on symlinks that target outside the root if danglings are not allowed. We should maybe err even if danglings are allowed, to prevent user error?

Note that technically the API usually refers to it as "Input Root", but we use "Exec Root" throughout the codebase... Which means we likely want to stick to that naming, even if "Exec" isn't necessarily pertinent to "CAS".

Reply:
Doing in a separate CL is probably a better idea, but leave a TODO.

Reply:
added a TODO.

Reply:
ah, I've added it to a different `Lstat` call. Now both have `WithStack`.